src/Aspire.Hosting/Utils/IEnvironmentVariables.cs

```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Aspire;
+
+/// <summary>
+/// An abstraction over the current process's environment variables.
+/// </summary>
+internal interface IEnvironmentVariables
+{
+    /// <summary>
+    /// Gets the named environment variable's value, or <see langword="null"/>.
+    /// </summary>
+    /// <param name="variableName">The name of the environment variable.</param>
+    /// <param name="defaultValue">An optional default value to return if the environment variable is not present.</param>
+    /// <returns>The named environment variable's value if present, or <see langword="null"/> if the variable is not present and no default was specified.</returns>
+    [return: NotNullIfNotNull(nameof(defaultValue))]
+    string? GetString(string variableName, string? defaultValue = null);
+}
+
+internal sealed class EnvironmentVariables : IEnvironmentVariables
+{
+    /// <summary>
+    /// A cache of all queried environment variables.
+    /// </summary>
+    private ImmutableDictionary<string, string?> _valueByName = ImmutableDictionary<string, string?>.Empty.WithComparers(StringComparers.EnvironmentVariableName);
+
+    [return: NotNullIfNotNull(nameof(defaultValue))]
+    public string? GetString(string variableName, string? defaultValue = null)
+    {
+        var value = ImmutableInterlocked.GetOrAdd(ref _valueByName, key: variableName, valueFactory: Environment.GetEnvironmentVariable);
```

Comment:
Why do we  need any caching at all? Why not just pass through to Environment.GetEnvironmentVariable?

Reply:
Looks like it allocates and P/Invokes each time:

https://github.com/dotnet/runtime/blob/d56fad46e213f8eb8d36be8fe7d342700c0a253c/src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs#L13-L31

I don't know if we care, but we do check env vars a fair bit around the place, and caching takes two lines of code.